src/library.js

```diff
@@ -588,9 +588,11 @@ mergeInto(LibraryManager.library, {
   $withStackSave__internal: true,
   $withStackSave: function(f) {
     var stack = stackSave();
-    var ret = f();
-    stackRestore(stack);
-    return ret;
+    try {
+      return f();
+    } finally {
+      stackRestore(stack);
+    }
```

Comment:
Is this necessary with this PR? It might be worth doing as a separate change if so. (there might be performance implications here)

Reply:
Yes, because a user callback attached to a promise may throw, resulting in a rejected callback. We should be able to handle that gracefully and without corrupting the stack pointer. The alternative to this change would be doing the same thing in a one-off specialized utility in the library_promise.js code.

I suppose I should probably add tests for this as well.

Reply:
I no longer modify or use withStackSave in the latest version.